File: internal/muter/muter.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// HUSHLINE — muter support utilities

package muter

import (
	"fmt"
	"regexp"

	"github.com/local/hushline/internal/config"
)

type QuietPattern struct {
	Name        string
	Pattern     *regexp.Regexp
	Replacement string
}

type Muter struct {
	rules        []QuietPattern
	maxLineWidth int
}

func Compose(cfg config.Config) (*Muter, error) {
	r := Muter{
		maxLineWidth: cfg.MaxLineWidth,
	}

	rules := make([]QuietPattern, 0, 2+len(cfg.MaskPatterns)+len(cfg.QuietRules))
	if cfg.StripANSI {
		rules = append(rules, QuietPattern{
			Name:        "ansi",
			Pattern:     ansiRe,
			Replacement: "",
		})
	}

	for _, pattern := range cfg.MaskPatterns {
		expr, err := regexp.Compile(pattern)
		if err != nil {
			return nil, fmt.Errorf("invalid redact pattern %q: %w", pattern, err)
		}
		rules = append(rules, QuietPattern{
			Name:        "redact",
			Pattern:     expr,
			Replacement: "***",
		})
	}

	for _, rr := range cfg.QuietRules {
		expr, err := regexp.Compile(rr.Pattern)
		if err != nil {
			return nil, fmt.Errorf("invalid silence pattern %q: %w", rr.Pattern, err)
		}
		rules = append(rules, QuietPattern{
			Name:        rr.Name,
			Pattern:     expr,
			Replacement: rr.Replacement,
		})
	}
	r.rules = rules
	return &r, nil
}

func (f *Muter) Apply(input string) string {
	if f == nil {
		return input
	}
	out := input
	for _, r := range f.rules {
		out = r.Pattern.ReplaceAllString(out, r.Replacement)
	}
	// ansi already included as a first rule but preserve clarity for intent
	if f.maxLineWidth > 0 {
		out = truncate(out, f.maxLineWidth)
	}
	return out
}

var ansiRe = regexp.MustCompile(`\x1b\[[0-9;]*[a-zA-Z]`)

func truncate(input string, max int) string {
	if max <= 0 {
		return input
	}
	if len(input) <= max {
		return input
	}
	return input[:max]
}

File: cores/core-python/src/hushline_core/muter.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# HUSHLINE — muter support utilities

"""Line shaping: ANSI stripping, secret redaction, and silence rewrites.

Rules run in the same order as the Go core: ANSI escape removal first (when
enabled), then each mask pattern replaced with ``***``, then the configured
silence rules. Finally, if a positive line width is set, the line is truncated
to that many bytes — matching Go's byte-slice truncation.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from hushline_core.config import Config

_ANSI_RE = re.compile(r"\x1b\[[0-9;]*[a-zA-Z]")
_MASK_REPLACEMENT = "***"


class MuterError(Exception):
    """Raised when a configured pattern is not a valid regular expression."""


@dataclass(frozen=True)
class _Rule:
    name: str
    pattern: re.Pattern[str]
    replacement: str


class Muter:
    """Applies the composed rule chain to a single line of text."""

    def __init__(self, rules: list[_Rule], max_line_width: int) -> None:
        self._rules = rules
        self._max_line_width = max_line_width

    def apply(self, text: str) -> str:
        out = text
        for rule in self._rules:
            out = rule.pattern.sub(rule.replacement, out)
        if self._max_line_width > 0:
            out = _truncate(out, self._max_line_width)
        return out


def compose(cfg: Config) -> Muter:
    """Build a :class:`Muter` from a resolved profile.

    Raises :class:`MuterError` if any mask or silence pattern fails to compile,
    mirroring the Go core's refusal to run with an invalid profile.
    """
    rules: list[_Rule] = []

    if cfg.strip_ansi:
        rules.append(_Rule(name="ansi", pattern=_ANSI_RE, replacement=""))

    for pattern in cfg.mask_patterns:
        rules.append(
            _Rule(
                name="redact",
                pattern=_compile(pattern, "invalid redact pattern"),
                replacement=_MASK_REPLACEMENT,
            )
        )

    for rule in cfg.quiet_rules:
        rules.append(
            _Rule(
                name=rule.name,
                pattern=_compile(rule.pattern, "invalid silence pattern"),
                replacement=rule.replacement,
            )
        )

    return Muter(rules, cfg.max_line_width)


def _compile(pattern: str, label: str) -> re.Pattern[str]:
    try:
        return re.compile(pattern)
    except re.error as exc:
        raise MuterError(f"{label} {pattern!r}: {exc}") from exc


def _truncate(text: str, max_bytes: int) -> str:
    """Truncate ``text`` to at most ``max_bytes`` UTF-8 bytes.

    Go slices the raw byte buffer (``input[:max]``); we bound by byte length the
    same way but decode back with ``errors="ignore"`` so a cut that lands inside
    a multi-byte rune never emits invalid UTF-8. Only called with a positive
    ``max_bytes`` (the caller guards on a positive line width).
    """
    encoded = text.encode("utf-8")
    if len(encoded) <= max_bytes:
        return text
    return encoded[:max_bytes].decode("utf-8", errors="ignore")
